ion-schema/src/constraint.rs

```diff
@@ -1079,66 +1079,44 @@ impl ConstraintValidator for FieldNamesConstraint {
-                    let schema_element: IonSchemaElement = (&Element::symbol(field_name)).into();
+        for (field_name, _) in ion_struct.iter() {
+            ion_path.push(IonPathElement::Field(field_name.text().unwrap().to_owned()));
+            let schema_element: IonSchemaElement = (&Element::symbol(field_name)).into();
 
-                    if let Err(violation) =
-                        self.type_reference
-                            .validate(&schema_element, type_store, ion_path)
-                    {
-                        violations.push(violation);
-                    }
-                    if self.requires_distinct && !field_name_set.insert(field_name.text().unwrap())
-                    {
-                        violations.push(Violation::new(
-                            "field_names",
-                            ViolationCode::ElementNotDistinct,
-                            format!("expected distinct field names but found duplicate field name {field_name}", ),
-                            ion_path,
-                        ))
-                    }
-                    ion_path.pop();
-                }
-                if !violations.is_empty() {
-                    return Err(Violation::with_violations(
-                        "field_names",
-                        ViolationCode::ElementMismatched,
-                        "one or more field names don't satisfy field_names constraint",
-                        ion_path,
-                        violations,
-                    ));
-                }
-                Ok(())
+            if let Err(violation) =
+                self.type_reference
+                    .validate(&schema_element, type_store, ion_path)
+            {
+                violations.push(violation);
             }
-            IonSchemaElement::SingleElement(element) => {
-                // return Violation if value is not a struct
-                Err(Violation::new(
+            if self.requires_distinct && !field_name_set.insert(field_name.text().unwrap()) {
+                violations.push(Violation::new(
                     "field_names",
-                    ViolationCode::TypeMismatched,
-                    format!("expected a struct but found {}", element.ion_type()),
+                    ViolationCode::ElementNotDistinct,
+                    format!(
+                        "expected distinct field names but found duplicate field name {field_name}",
+                    ),
                     ion_path,
                 ))
             }
-            IonSchemaElement::Document(_) => Err(Violation::new(
+            ion_path.pop();
+        }
+
+        if !violations.is_empty() {
+            return Err(Violation::with_violations(
                 "field_names",
-                ViolationCode::TypeMismatched,
-                "expected a struct but found document",
+                ViolationCode::ElementMismatched,
```

Comment:
Doesn't need to be this PR, but we should evaluate whether `ElementMismatched` and `ElementNotDistinct` are suitable to re-use for the `field_names` constraint.